src/ucs/sys/socket.c

```diff
@@ -0,0 +1,193 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#include <ucs/debug/log.h>
+#include <ucs/sys/string.h>
+#include <ucs/sys/socket.h>
+#include <ucs/sys/math.h>
+
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+
+
+ucs_status_t ucs_netif_ioctl(const char *if_name, unsigned long request,
+                             struct ifreq *if_req)
+{
+    ucs_status_t status;
+    int fd, ret;
+
+    ucs_strncpy_zero(if_req->ifr_name, if_name, sizeof(if_req->ifr_name));
+
+    status = ucs_socket_create(AF_INET, SOCK_STREAM, &fd);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    ret = ioctl(fd, request, if_req);
+    if (ret < 0) {
+        ucs_debug("ioctl(req=%lu, ifr_name=%s) failed: %m", request, if_name);
+        status = UCS_ERR_IO_ERROR;
+        goto out_close_fd;
+    }
+
+    status = UCS_OK;
+
+out_close_fd:
+    close(fd);
+out:
+    return status;
+}
+
+int ucs_netif_is_active(const char *if_name)
+{
+    ucs_status_t status;
+    struct ifreq ifr;
+
+    status = ucs_netif_ioctl(if_name, SIOCGIFADDR, &ifr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    status = ucs_netif_ioctl(if_name, SIOCGIFFLAGS, &ifr);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    return (ifr.ifr_flags & IFF_UP) && (ifr.ifr_flags & IFF_RUNNING) &&
+           !(ifr.ifr_flags & IFF_LOOPBACK);
+}
+
+ucs_status_t ucs_socket_create(int domain, int type, int *fd_p)
+{
+    int fd = socket(domain, type, 0);
+    if (fd < 0) {
+        ucs_error("socket create failed: %m");
+        return UCS_ERR_IO_ERROR;
+    }
+
+    *fd_p = fd;
+    return UCS_OK;
+}
+
+ucs_status_t ucs_socket_connect(int fd, const struct sockaddr *dest_addr)
+{
+    char str[UCS_SOCKADDR_STRING_LEN];
+    int ret;
+
+    ret = connect(fd, dest_addr, ucs_sockaddr_sizeof(dest_addr));
```

Comment:
should we also handle EINTR here?

Reply:
good catch! we should handle it here

btw, should we handle `EISCONN` and return `UCS_OK`?

Reply:
i think we should fail in case of EISCONN, since it means we are calling connect() twice on same socket

Reply:
@yosefe do we need to return `UCS_ERR_ALREADY_EXISTS`?

Reply:
yes

Reply:
done